### PR TITLE
Python API for storing and loading chat history

### PR DESCRIPTION
**Pull request by o01eg**
Introduce dummy implementation for storing and loading chat history via python script which could be altered to use an actual DBMS or so.

Introduce class `ChatProvider` implements methods `load_history` to read chat history from external source and `put_history_entity` to store it.

P.S. Example for psycopg2:
```python
from logging import info

from common.configure_logging import redirect_logging_to_freeorion_logger

# Logging is redirected before other imports so that import errors appear in log files.
redirect_logging_to_freeorion_logger()

import freeorion as fo

import psycopg2
import psycopg2.extensions
psycopg2.extensions.register_type(psycopg2.extensions.UNICODE)
psycopg2.extensions.register_type(psycopg2.extensions.UNICODEARRAY)


class ChatHistoryProvider:
    def __init__(self):
        """
        Initializes ChatProvider. Doesn't accept arguments.
        """
        dsn = ""
        with open(fo.get_user_config_dir() + "/db.txt", "r") as f:
            for line in f:
                dsn = line
                break
        self.conn = psycopg2.connect(dsn)
        info("Chat initialized")

    def load_history(self):
        """
        Loads chat history from external sources.

        Returns list of tuples:
            (unixtime timestamp, player name, text, text color of type freeorion.GGColor)
        """
        info("Loading history...")
        res = []
        with self.conn:
            with self.conn.cursor() as curs:
                curs.execute(""" SELECT date_part('epoch', ts)::int, player_name, text,
                    text_color / 256 / 256 / 256 % 256,
                    text_color / 256 / 256 % 256,
                    text_color / 256 % 256,
                    text_color % 256
                    FROM chat_history
                    ORDER BY ts
                    LIMIT 1000 """)
                for r in curs:
                    c = fo.GGColor(r[3], r[4], r[5], r[6])
                    e = (r[0], str(r[1].encode('utf-8')), str(r[2].encode('utf-8')), c)
                    res.append(e)
        return res

    def put_history_entity(self, timestamp, player_name, text, text_color):
        """
        Put chat into external storage.

        Return True if successfully stored. False otherwise.

        Arguments:
        timestamp -- unixtime, number of seconds from 1970-01-01 00:00:00 UTC
        player_name -- player name
        text -- chat text
        text_color -- freeorion.GGColor
        """
        info("Chat %s: %s %s" % (player_name, text, text_color))
        with self.conn:
            with self.conn.cursor() as curs:
                curs.execute(""" INSERT INTO chat_history (ts, player_name, text, text_color)
                    VALUES (to_timestamp(%s) at time zone 'utc', %s, %s, %s)""",
                    (timestamp, player_name, text,
                        256 * (256 * (256 * text_color.r + text_color.g) + text_color.b) + text_color.a))
        return True
```

### CODE REVIEW

**Comment by Cjkjvfnby** [comment]
Why is it important?

File: default/python/chat/chat.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+        info("Chat initialized")
+
+    def load_history(self):
+        info("Loading history...")
+        # c = fo.GGColor(255, 128, 128, 255)
+        # e = (123456789012, "P1", "Test1", c)
+        # return [e]
+        return []
+
+    def put_history_entity(self, timestamp, player_name, text, text_color):
```
Comment (Cjkjvfnby) [inline default/python/chat/chat.py:29]: Add docstring.

**Review by Cjkjvfnby — CHANGES_REQUESTED** [review]
Can we store chat history in savegame?

File: default/python/chat/chat.py (hunk 1 above)
Comment (o01eg) [reply default/python/chat/chat.py:29]: Added

File: default/python/chat/chat.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+# Logging is redirected before other imports so that import errors appear in log files.
+redirect_logging_to_freeorion_logger()
+
+
+class ChatProvider:
+    def __init__(self):
+        info("Chat initialized")
+
+    def load_history(self):
+        info("Loading history...")
```
[the remark this answers is not in the document]
Comment (o01eg) [reply default/python/chat/chat.py]: Added

**Comment by o01eg** [comment]
> Why is it important?

Players use chat to arrange next games. Currently the server doesn't show them anything they said in chat after restarts or failures.

> Can we store chat history in savegame?

It mostly used in lobby where savegames aren't available.

File: default/python/chat/chat.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
+from logging import info
+
+from common.configure_logging import redirect_logging_to_freeorion_logger
+
+# Logging is redirected before other imports so that import errors appear in log files.
+redirect_logging_to_freeorion_logger()
+
+
+class ChatProvider:
```
Comment (geoffthemedio) [inline default/python/chat/chat.py]: What exactly is this providing, or will it provide in general? Just chat history? If so, maybe name it more specifically?

Comment (o01eg) [reply default/python/chat/chat.py]: Will `ChatHistoryProvider` be better name?

Comment (o01eg) [reply default/python/chat/chat.py]: Changed.

**Comment by Cjkjvfnby** [comment]
> Players use chat to arrange next games. Currently the server doesn't show them anything they said in chat after restarts or failures.

I am not familiar with this chat at all.  `chat history` is the ambiguous term for me, it references to lobby chat and game chat.   Please add `lobby` prefix (or some other)  to all documentation. 

Can we just have one provider, that will store this history in the file in the logs directory  (or maybe in multiple files in the subdirectory, to have good pagination)?  This should be the game feature that not require external tools. 

PS Which platform do you use?  On Windows python shipped with the game and it fails on my machine if I try to use some external libraries (as I recall all code that use _sockets.py affected). I expect that `psycopg2` will not work for me.

**Comment by o01eg** [comment]
> I am not familiar with this chat at all. chat history is the ambiguous term for me, it references to lobby chat and game chat. Please add lobby prefix (or some other) to all documentation.

Chat is common for lobby and game.

> Can we just have one provider, that will store this history in the file in the logs directory (or maybe in multiple files in the subdirectory, to have good pagination)? This should be the game feature that not require external tools.

I'm not sure how to manage and clean up multiple files. Also I'm not sure if this storage is really needed for regular games.

Dummy implementation doesn't require any external tools.

> PS Which platform do you use? On Windows python shipped with the game and it fails on my machine if I try to use some external libraries (as I recall all code that use _sockets.py affected). I expect that `psycopg2` will not work for me.

I use Linux. Script with `psycopg2`  is only example how it can be altered. It's not for shipping with the game.

**Comment by geoffthemedio** [comment]
Is the current version of his supposed to actually output a db.txt and restore it when restarted? I don't get such a file in my FreeOrion user settings directory, and chat isn't restored and resent when starting a new local server.

**Comment by o01eg** [comment]
No, `db.txt` is a file for an example how it could be altered. "Current version" in branch for merge don't require neither`db.txt` file nor `psycopg2` package. Actually it doesn't load and store as well.

**Pull request merged**